File: ios-coverage/CoverageCollector.m
//
//  CoverageCollector.m
//  iOS Code Coverage Collector
//

#import "CoverageCollector.h"

// LLVM Profile 运行时函数声明
#ifndef PROFILE_INSTRPROFILING_H_
#define PROFILE_INSTRPROFILING_H_

extern int __llvm_profile_runtime;
void __llvm_profile_initialize_file(void);
const char *__llvm_profile_get_filename(void);
void __llvm_profile_set_filename(const char *);
int __llvm_profile_write_file(void);
int __llvm_profile_register_write_file_atexit(void);
const char *__llvm_profile_get_path_prefix(void);

#endif /* PROFILE_INSTRPROFILING_H_ */

static NSString *s_coverageDirectory = nil;
static NSString *s_coverageFilePath = nil;

@implementation CoverageCollector

+ (void)initializeCollector {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSString *coveragePath = [self coverageFilePath];
        if (coveragePath) {
            __llvm_profile_set_filename([coveragePath UTF8String]);
            NSLog(@"[Coverage] Initialized with output path: %@", coveragePath);
        }
    });
}

+ (BOOL)dumpCoverageData {
    int result = __llvm_profile_write_file();
    if (result == 0) {
        NSLog(@"[Coverage] Data saved successfully to: %@", [self coverageFilePath]);
        return YES;
    } else {
        NSLog(@"[Coverage] Failed to save data, error code: %d", result);
        return NO;
    }
}

+ (nullable NSString *)coverageFilePath {
    // 固定路径：第一次调用时算好就缓存住，后续调用（dumpCoverageData 的日志、CoverageUploader
    // 取最新文件）必须拿到同一个路径，否则会拿到一个从未被 __llvm_profile_write_file 写过的新路径
    if (s_coverageFilePath) {
        return s_coverageFilePath;
    }
    NSString *directory = [self coverageDirectory];
    NSString *bundleIdentifier = [[NSBundle mainBundle] bundleIdentifier] ?: @"unknown";
    NSString *timestamp = [self currentTimestamp];
    NSString *fileName = [NSString stringWithFormat:@"%@_%@.profraw", bundleIdentifier, timestamp];
    s_coverageFilePath = [directory stringByAppendingPathComponent:fileName];
    return s_coverageFilePath;
}

+ (NSArray<NSString *> *)allCoverageFiles {
    NSString *directory = [self coverageDirectory];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSError *error = nil;
    
    NSArray *files = [fileManager contentsOfDirectoryAtPath:directory error:&error];
    if (error) {
        NSLog(@"[Coverage] Error reading directory: %@", error);
        return @[];
    }
    
    NSMutableArray *coverageFiles = [NSMutableArray array];
    for (NSString *file in files) {
        if ([file hasSuffix:@".profraw"]) {
            [coverageFiles addObject:[directory stringByAppendingPathComponent:file]];
        }
    }
    
    return [coverageFiles copy];
}

+ (void)clearCoverageData {
    NSString *directory = [self coverageDirectory];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSError *error = nil;
    
    NSArray *files = [fileManager contentsOfDirectoryAtPath:directory error:&error];
    for (NSString *file in files) {
        if ([file hasSuffix:@".profraw"]) {
            NSString *filePath = [directory stringByAppendingPathComponent:file];
            [fileManager removeItemAtPath:filePath error:nil];
        }
    }
    
    NSLog(@"[Coverage] All coverage data cleared");
}

+ (void)setCoverageDirectory:(NSString *)directory {
    s_coverageDirectory = [directory copy];
    s_coverageFilePath = nil; // 目录变了，让 coverageFilePath 下次调用时按新目录重新计算

    // 确保目录存在
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath:directory]) {
        NSError *error = nil;
        [fileManager createDirectoryAtPath:directory
               withIntermediateDirectories:YES
                                attributes:nil
                                     error:&error];
        if (error) {
            NSLog(@"[Coverage] Failed to create directory: %@", error);
        }
    }
}

+ (NSString *)coverageDirectory {
    if (s_coverageDirectory) {
        return s_coverageDirectory;
    }
    
    // 默认使用 Document 目录下的 Coverage 子目录
    NSArray *paths = NSSearchPathForDirectoriesInDomains(
        NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths firstObject];
    NSString *coverageDir = [documentsDirectory stringByAppendingPathComponent:@"Coverage"];
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath:coverageDir]) {
        [fileManager createDirectoryAtPath:coverageDir
               withIntermediateDirectories:YES
                                attributes:nil
                                     error:nil];
    }
    
    return coverageDir;
}

+ (NSString *)currentTimestamp {
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyyMMdd_HHmmss"];
    return [formatter stringFromDate:[NSDate date]];
}

@end

File: ios-coverage/CoverageCollector.swift
//
//  CoverageCollector.swift
//  iOS Code Coverage Collector (Swift 版本)
//

import Foundation

/// iOS 代码覆盖率收集器
public class CoverageCollector {
    
    // MARK: - Properties
    
    private static var coverageDirectory: URL?
    
    // LLVM Profile 函数声明
    private static let llvmProfileWriteFile: @convention(c) () -> Int32 = {
        let handle = dlopen(nil, RTLD_NOW)
        let sym = dlsym(handle, "__llvm_profile_write_file")
        return unsafeBitCast(sym, to: (@convention(c) () -> Int32).self)
    }()
    
    private static let llvmProfileSetFilename: @convention(c) (UnsafePointer<CChar>) -> Void = {
        let handle = dlopen(nil, RTLD_NOW)
        let sym = dlsym(handle, "__llvm_profile_set_filename")
        return unsafeBitCast(sym, to: (@convention(c) (UnsafePointer<CChar>) -> Void).self)
    }()
    
    // MARK: - Public Methods
    
    /// 初始化覆盖率收集器
    public static func initialize() {
        DispatchQueue.once {
            if let path = coverageFilePath()?.path {
                llvmProfileSetFilename(path)
                print("[Coverage] Initialized with output path: \(path)")
            }
        }
    }
    
    /// 保存覆盖率数据
    @discardableResult
    public static func dumpCoverageData() -> Bool {
        let result = llvmProfileWriteFile()
        if result == 0 {
            print("[Coverage] Data saved successfully")
            return true
        } else {
            print("[Coverage] Failed to save data, error code: \(result)")
            return false
        }
    }
    
    /// 获取覆盖率文件路径
    public static func coverageFilePath() -> URL? {
        let directory = coverageDirectory ?? defaultCoverageDirectory()
        let bundleIdentifier = Bundle.main.bundleIdentifier ?? "unknown"
        let timestamp = currentTimestamp()
        let fileName = "\(bundleIdentifier)_\(timestamp).profraw"
        return directory.appendingPathComponent(fileName)
    }
    
    /// 获取所有覆盖率文件
    public static func allCoverageFiles() -> [URL] {
        let directory = coverageDirectory ?? defaultCoverageDirectory()
        
        do {
            let files = try FileManager.default.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: nil
            )
            return files.filter { $0.pathExtension == "profraw" }
        } catch {
            print("[Coverage] Error reading directory: \(error)")
            return []
        }
    }
    
    /// 清除所有覆盖率数据
    public static func clearCoverageData() {
        let files = allCoverageFiles()
        for file in files {
            try? FileManager.default.removeItem(at: file)
        }
        print("[Coverage] All coverage data cleared")
    }
    
    /// 设置覆盖率输出目录
    public static func setCoverageDirectory(_ directory: URL) {
        coverageDirectory = directory
        
        // 确保目录存在
        try? FileManager.default.createDirectory(
            at: directory,
            withIntermediateDirectories: true
        )
    }
    
    // MARK: - Private Methods
    
    private static func defaultCoverageDirectory() -> URL {
        let documentsPath = FileManager.default.urls(
            for: .documentDirectory,
            in: .userDomainMask
        ).first!
        
        let coveragePath = documentsPath.appendingPathComponent("Coverage")
        
        try? FileManager.default.createDirectory(
            at: coveragePath,
            withIntermediateDirectories: true
        )
        
        return coveragePath
    }
    
    private static func currentTimestamp() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        return formatter.string(from: Date())
    }
}

// MARK: - DispatchQueue Extension

extension DispatchQueue {
    private static var _onceTracker = [String]()
    
    public class func once(file: String = #file, function: String = #function, line: Int = #line, block: () -> Void) {
        let token = file + ":" + function + ":" + String(line)
        once(token: token, block: block)
    }
    
    public class func once(token: String, block: () -> Void) {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        
        if _onceTracker.contains(token) {
            return
        }
        
        _onceTracker.append(token)
        block()
    }
}

// MARK: - Usage in AppDelegate

/*
import UIKit

@main
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    func application(_ application: UIApplication, 
                     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // 初始化覆盖率收集器
        CoverageCollector.initialize()
        return true
    }
    
    func applicationDidEnterBackground(_ application: UIApplication) {
        // App 进入后台时保存覆盖率数据
        CoverageCollector.dumpCoverageData()
    }
    
    func applicationWillTerminate(_ application: UIApplication) {
        // App 即将终止时保存覆盖率数据
        CoverageCollector.dumpCoverageData()
    }
}
*/
